csharp/src/Apache.Arrow/Arrays/ArrayDataConcatenator.cs

```diff
@@ -71,7 +71,7 @@ public ArrayDataConcatenationVisitor(IReadOnlyList<ArrayData> arrayDataList, Mem
                 foreach (ArrayData arrayData in _arrayDataList)
                 {
                     _totalLength += arrayData.Length;
-                    _totalNullCount += arrayData.NullCount;
+                    _totalNullCount += arrayData.GetNullCount();
```

Comment:
It could make sense to check `NullCount` for -1 here, and then set the final null count to -1 if it was -1 in any array, rather than forcing a computation. But this class has bigger problems as it doesn't account for non-zero offsets anywhere as far as I can see, although it does seem to account for array lengths being less than the full buffer size. Maybe it should throw a `NotImplementedException` if it encounters a non-zero offset?

Reply:
Can you file a separate bug for this?

Reply:
Yep sure, I've made #41164